### PR TITLE
Handle to_s for amounts larger than float precision

### PR DESCRIPTION
**Pull request by tgwizard**
Fix https://github.com/Shopify/money/issues/211.

The crux of the issue is that some numbers, like `99999999999999.91`, `99999999999999.99`, etc., cannot be represented as floats:

```
irb(main):001:0> 99999999999999.91
=> 99999999999999.9
irb(main):002:0> 99999999999999.99
=> 99999999999999.98
```

The old `to_s` method used the `f` field for `sprintf`, which essentially converts the `Money` amount `BigDecimal` to a float, then to a string.

This PR changes the implementation of `to_s` so that we don't use float as an intermediate representation. Unfortunately, it's a bit messy. It's inspired by https://stackoverflow.com/a/24892562. Feedback and suggestions for improvement more than welcome.

### CODE REVIEW

**Comment by tgwizard** [comment]
CI seems to fail not because of tests, but because of coverage collection.

File: lib/money/money.rb [hunk 1]
```diff
@@ -207,11 +207,20 @@ def to_d
...
-    case style
+    units = case style
     when :legacy_dollars
-      sprintf("%.2f", value)
+      2
     when :amount, nil
-      sprintf("%.#{currency.minor_units}f", value)
+      currency.minor_units
+    else
+      return nil
```
Comment (elfassy) [inline lib/money/money.rb:216]: any reason we would not want to raise here? We expect style to be one of `:legacy_dollars`, `:amount`, or nil.

**Comment by tgwizard** [comment]
@elfassy you're right, `99999999999999.91` as a string/BigDecimal behaves correctly on the master version of Shopify/money, with `to_s` using `%.2f` and all. I used it as an example in the PR description only of something that can't be represented as a float:

```
[3] pry(main)> BigDecimal('99999999999999.91').to_f
=> 99999999999999.9
```

The other example from the PR description, as well as the example for the original issue, behave incorrectly, with rounding errors also with `sprintf`, `%.2f`, and here in Shopify/money.

**Comment by elfassy** [comment]
Looks to me like there's some issue with the precision we're using in the Money::Helper class
```
[35] pry(main)> BigDecimal( 99999999999999.99, 16)
=> 0.9999999999999998e14
[36] pry(main)> Float::DIG
=> 15
[37] pry(main)> BigDecimal( 99999999999999.99, 15)
=> 0.1e15
[38] pry(main)> BigDecimal( 99999999999999.99, 17)
ArgumentError: precision too large.
from (pry):38:in `BigDecimal'
```
For some reason `Float::DIG` is 15, but we're allowed to use a precision of 16. This will at least keep the result consistent between a strong and a float

File: lib/money/money.rb [hunk 2]
```diff
@@ -207,11 +207,20 @@ def to_d
...
+    end
+
+    rounded = round(units)
```
Comment (elfassy) [inline lib/money/money.rb:219]: the `value` is already rounded. How about
```ruby
def to_s(style = nil)
  units = (style == :legacy_dollars ? 2 : currency.minor_units)
  if units == 0
    sprintf("%d", value)
  else
    sprintf("%d.%0#{units}d", value.truncate, value.frac * (10 ** units))
  end
end
```

Comment (tgwizard) [reply lib/money/money.rb:219]: The value isn't always already rounded. Example is `Money.new("1.999", "USD)` (or possibly `Money.new("1.99", "USD) / 2`. These amounts have 3 decimals, and on master they are rendered as `"2.00"` and `"1.00"`, respectively. We need to round for that.

`units = (style == :legacy_dollars ? 2 : currency.minor_units)` is slightly breaking. The old version returns `nil` if `style` is an unexpected value like `:foo`. That's why my code looks like it does. I'm happy to change to what you propose if we're fine with the breaking change.

File: lib/money/money.rb (hunk 1 above)
Comment (elfassy) [inline lib/money/money.rb:216]: >CI seems to fail not because of tests, but because of coverage collection.

I believe this line of code is not tested. Code coverage needs to be 100% on this repo

Comment (tgwizard) [reply lib/money/money.rb:216]: This line of code existed before (case/switch statement without an `else` clause), but depending on discussion in https://github.com/Shopify/money/pull/212#discussion_r504044321, i'm happy to add a test that asserts an unknown `style` returns `nil`.

File: lib/money/money.rb (hunk 2 above)
Comment (elfassy) [reply lib/money/money.rb:219]: sorry i meant the value is rounded when initialized here https://github.com/Shopify/money/blob/master/lib/money/money.rb#L87. You're right though that this won't work for a 3 digit currency we're trying to display using "legacy_dollars" notation. 

I belive you can remove the `to_d` here since these would already all be decimals

Comment (tgwizard) [reply lib/money/money.rb:219]: Right, I can change `rounded = round(units)` to `rounded_value = @value.round(units)`, and skip the `to_d` parts.

File: lib/money/money.rb (hunk 1 above)
Comment (elfassy) [reply lib/money/money.rb:216]: Looks like we were missing a test, probably best to make it raise an argument error if it's not something we expect

File: spec/money_spec.rb [hunk 3]
```diff
@@ -80,6 +80,23 @@
...
+
+  end
+
+  it "to_s does not round when fractions same as currency allows" do
+    expect(Money.new("1.25", "USD").to_s).to eq("1.25")
+    expect(Money.new("9.99", "USD").to_s).to eq("9.99")
+    expect(Money.new("9.999", "BHD").to_s).to eq("9.999")
+  end
+
+  it "to_s does not found if amount is larger than float allows" do
```
Comment (bdewater) [inline spec/money_spec.rb:95]: found -> round :)

**Comment by tgwizard** [comment]
@elfassy @bdewater I've addressed your comments. Please have another look. (I plan on squashing when merging in the end).

**Review by elfassy — APPROVED** [review]

File: spec/money_spec.rb [hunk 4]
```diff
@@ -80,6 +80,26 @@
...
+
+  it "to_s does not round when fractions same as currency allows" do
+    expect(Money.new("1.25", "USD").to_s).to eq("1.25")
+    expect(Money.new("9.99", "USD").to_s).to eq("9.99")
+    expect(Money.new("9.999", "BHD").to_s).to eq("9.999")
+  end
+
+  it "to_s does not round if amount is larger than float allows" do
+    expect(Money.new("99999999999999.99", "USD").to_s).to eq("99999999999999.99")
+    expect(Money.new("999999999999999999.99", "USD").to_s).to eq("999999999999999999.99")
```
Comment (elfassy) [inline spec/money_spec.rb:96]: please also add `99999999999999.99` decimal use case (will require you to change the precision from 15 to 16)

Comment (tgwizard) [reply spec/money_spec.rb:96]: hmm, you mean 

```ruby
expect(Money.new(BigDecimal( 99999999999999.99, 16), "USD").to_s).to eq("999999999999999999.99")
```

or change `Money::Helpers#value_to_decimal` to use `BigDecimal(num, 16)` instead of `BigDecimal(num, Float::DIG)`?

**Review by bdewater — APPROVED** [review]

File: spec/money_spec.rb (hunk 4 above)
Comment (elfassy) [reply spec/money_spec.rb:96]: just adding, to catch the case where this was changed to 10000...
```ruby
expect(Money.new(99999999999999.99, "USD").to_s).to eq("99999999999999.99")
```
